lib/compress/zstd_fast.c

```diff
@@ -183,6 +183,237 @@ ZSTD_compressBlock_fast_generic(
 }
 
 
+/**
+ * If you squint hard enough (and ignore repcodes), the search operation at any
+ * given position is broken into 4 stages:
+ *
+ * 1. Hash   (map position to hash value via input read)
+ * 2. Lookup (map hash val to index via hashtable read)
+ * 3. Load   (map index to value at that position via input read)
```

Comment:
I was expecting a sort of `prefetch` between `lookup` and `load`. No need ?

Reply:
My sense is that by the time the lookup result comes back, letting us figure out where we'd issue the prefetch, we're already executing the load, I think. The core loop is only 34 instructions.

And other experimentation shows that the match load is not a significant bottleneck (which was surprising to me). I guess that for fast, the match addresses tend to be nearby and already hot.

Which is what led me to actually collapse the distance between issuing the match load and operating on that result (the match compare). They're now just adjacent operations in the same loop iteration. Which turns out to be faster. Somehow. I guess it's worth it to shorten the lookahead so less work is lost / re-done at each taken match.

Reply:
Since this analysis seems dependent on a question of distance, 
could this conclusion differ depending on level 1 or level 2 ?

They both use `zstd_fast` strategy, but level 2 uses a lot more memory, hence is likely to fit into L1 cache.

Reply:
Potentially. It looks like the current version of the PR is a win on level 2, but maybe less so than on level 1.

But broadly I think my point stands: the pressure of not doing too much work ahead of time that will be wasted when we find a match means we don't want the pipeline to be too long. And so there isn't really any gap between the hashtable lookup and the load in which to issue a prefetch. It's pretty dense.

Experimentally, throwing a prefetch in the current loop is -4% speed.

Side note: Sunny Cove cores have 48KB L1! I wish I could get my hands on one of those. :)